### PR TITLE
feat: add optional MLflow experiment tracking support

### LINKED ISSUE #343: [Feature]: MLflow experiment tracking integration
## Problem

Experiment tracking is currently file-based (JSON results in output directories). There's no centralized experiment registry, no comparison UI beyond the custom frontend, and no standard way to query past runs. MLflow is the open standard for experiment tracking and integrates with most ML tools.

## Proposed Solution

1. Add MLflow logging as an optional feature (enabled via config flag)
2. Log: hyperparameters, metrics per cycle, final evaluation metrics, artifacts (model checkpoints)
3. Allow both local MLflow server and remote (Databricks, managed) backends
4. Pipeline phases should auto-log without user code changes

## Acceptance Criteria

- [ ] MLflow integration via `nightmarenet[mlflow]` optional extra
- [ ] Config flag: `tracking.backend: mlflow` enables MLflow logging
- [ ] Hyperparameters logged at run start
- [ ] Metrics logged per cycle (robustness score, loss, etc.)
- [ ] Final evaluation metrics logged as summary
- [ ] Model checkpoint logged as artifact
- [ ] Works with local MLflow server (no external service required for testing)
- [ ] Test verifies metrics appear in MLflow (using `mlflow.get_run()`)

## Scope

- `nightmarenet/tracking/mlflow_tracker.py` - new module
- `nightmarenet/tracking/__init__.py` - tracker factory (dispatches to mlflow or file-based)
- `pyproject.toml` - add `[mlflow]` optional extra
- `configs/examples/mlflow-tracking.yaml` - example config
- `tests/test_mlflow_tracker.py` - new test file

> **Important:** Please read CONTRIBUTING.md in full before requesting assignment. Assignment requests without a technical approach will not be considered. PRs demonstrating thorough research, benchmarks, or creative approaches beyond minimum criteria may earn bonus XP labels.

### PR DESCRIPTION
**Pull request by Ayush-0918**
## Summary

Add optional MLflow experiment tracking support by extending the existing `ExperimentTracker` abstraction while preserving compatibility with the existing WandB and TensorBoard integrations.

## Motivation

NightmareNet currently supports experiment tracking through WandB and TensorBoard but lacks native MLflow support. This change introduces MLflow as an optional backend, allowing users to log experiment parameters, metrics, and checkpoint artifacts without modifying the existing training workflow.

Closes #343

## Changes

- Added an optional `mlflow` dependency in `pyproject.toml`.
- Extended `nightmarenet/utils/tracking.py` to support the `mlflow` backend.
- Added MLflow initialization, parameter logging, metric logging, artifact logging, and run finalization.
- Added `log_artifact()` support to `ExperimentTracker`.
- Integrated checkpoint artifact logging in `nightmarenet/training/trainer.py`.
- Added `mlflow` as a supported tracking backend in:
  - `scripts/train.py`
  - `scripts/evaluate.py`
- Added example configuration:
  - `configs/examples/mlflow-tracking.yaml`
- Added tests for the MLflow tracker:
  - `tests/test_mlflow_tracker.py`

## Acceptance Criteria

- [x] Added optional MLflow experiment tracking support.
- [x] Hyperparameters are logged to MLflow.
- [x] Training metrics are logged during execution.
- [x] Checkpoints are logged as MLflow artifacts.
- [x] Existing WandB and TensorBoard functionality remains unchanged.
- [x] MLflow integration gracefully falls back when unavailable.
- [x] Added tests covering MLflow tracking functionality.
- [x] Added an example MLflow configuration.

## Type

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] Feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would break existing behavior)
- [ ] Refactor (no functional change)
- [ ] Documentation
- [x] Tests
- [ ] Infrastructure (CI/CD, Docker, tooling)

---

## Pre-submission Requirements

> These are mandatory. PRs missing any item will not be reviewed.

- [ ] I have **starred** this repository ⭐
- [ ] I have **followed** [@Adit-Jain-srm](https://github.com/Adit-Jain-srm)
- [x] I have read [CONTRIBUTING.md](https://github.com/Adit-Jain-srm/NightmareNet/blob/main/CONTRIBUTING.md) in full

## Quality Checklist

- [x] `ruff check nightmarenet/ scripts/ tests/` — 0 errors
- [ ] `mypy nightmarenet/ --ignore-missing-imports` — no new errors (run on Python 3.12)
- [ ] `pytest tests/` — all tests pass
- [x] New functionality has corresponding tests
- [x] Commit messages follow [Conventional Commits](https://www.conventionalcommits.org/) (`feat:`, `fix:`, `docs:`, `test:`, `refactor:`, `chore:`)
- [x] No `from __future__ import annotations` added under `nightmarenet/api/`
- [x] No new imports of hosted-only libraries (`sqlalchemy`, `redis`, `celery`) in `nightmarenet/`

## Documentation

- [ ] No documentation needed (test-only or internal refactor)
- [ ] README.md updated
- [x] Docstrings added/updated (Google style, public APIs only)
- [ ] `docs/` updated (architecture, research, or API docs)
- [ ] Config comments updated (`configs/default.yaml`)
- [ ] CHANGELOG entry added (if user-facing change)

## AI Disclosure

- [ ] This PR is entirely human-written
- [x] This PR includes AI-assisted code (Copilot, ChatGPT, Claude, etc.) — I have reviewed every line and can explain it

## Security Considerations

- [x] Not applicable (no security-sensitive changes)
- [ ] User input is validated before use
- [x] No secrets, keys, or credentials in code or comments
- [ ] CORS / auth / rate limiting behavior unchanged (or explicitly documented)
- [ ] If this is a security fix, see [SECURITY.md](https://github.com/Adit-Jain-srm/NightmareNet/blob/main/SECURITY.md) — do NOT describe the vulnerability publicly until patched

## Screenshots (if UI/UX change)

N/A

## Breaking Changes

N/A

---

<details>
<summary>Reviewer notes (optional)</summary>

This implementation extends the existing tracking abstraction instead of introducing a new tracking package, keeping the change backward compatible and minimizing the overall diff. MLflow support is optional and all MLflow operations are wrapped to avoid impacting training if the dependency is unavailable or tracking fails.

</details>

## Summary by CodeRabbit

- **New Features**
  - Added **MLflow** as an experiment-tracking option for training and evaluation, including metrics, flattened config parameters, and checkpoint/file/directory artifact logging.
  - Introduced a local MLflow example configuration.
  - Added an MLflow integration test; added the `tracking` extra dependency for MLflow.

- **Bug Fixes**
  - Checkpoint saving no longer fails if artifact logging errors (warning instead).
  - Improved AUC trapezoid fallback behavior when `np.trapezoid` isn’t available.

- **Tests**
  - Added coverage for MLflow runs, metrics, params, and artifact persistence.

### CODE REVIEW

File: tests/test_mlflow_tracker.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    runs = client.search_runs([experiment.experiment_id])
+    assert len(runs) == 1
+
+    run = runs[0]
+
+    assert run.data.params.get("training/learning_rate") == "0.001"
+    assert run.data.params.get("training/batch_size") == "32"
+
+    assert float(run.data.metrics.get("loss")) == 0.5
+    assert float(run.data.metrics.get("accuracy")) == 0.9
```
[the remark this answers is not in the document]
Comment (Ayush-0918) [reply tests/test_mlflow_tracker.py:55]: Thanks for the thoughtful review! I’ve implemented the suggested changes and verified them locally. Appreciate the feedback.

File: tests/test_mlflow_tracker.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    tracker = ExperimentTracker(
+        backend="mlflow",
+        project="TestMLflow",
+        config=config,
+    )
+
+    # Check if mlflow is installed (soft-skip if not)
+    if tracker.backend == "none":
+        pytest.skip("mlflow is not installed")
+
```
[the remark this answers is not in the document]
Comment (Ayush-0918) [reply tests/test_mlflow_tracker.py]: Thanks for the thoughtful review! I’ve implemented the suggested changes and verified them locally. Appreciate the feedback.

File: tests/test_mlflow_tracker.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import pytest
+
+from nightmarenet.utils.tracking import ExperimentTracker
+
+
+def test_mlflow_tracker_basic(tmp_path):
+    # Use local file backend for mlflow
+    mlruns_dir = tmp_path / "mlruns"
+    tracking_uri = f"file://{mlruns_dir}"
```
[the remark this answers is not in the document]
Comment (Ayush-0918) [reply tests/test_mlflow_tracker.py]: Thanks for the thoughtful review! I’ve implemented the suggested changes and verified them locally. Appreciate the feedback.

File: nightmarenet/utils/tracking.py [hunk 4]
```diff
@@ -152,6 +180,43 @@ def log_config(self, config: dict) -> None:
...
+
+                if os.path.isdir(path):
+                    mlflow.log_artifacts(path)
+                elif os.path.exists(path):
+                    mlflow.log_artifact(path)
+                else:
+                    logger.warning("Artifact path does not exist: %s", path)
+            except Exception as exc:
+                logger.warning("Failed to log artifact to MLflow: %s", exc)
+
```
[the remark this answers is not in the document]
Comment (Ayush-0918) [reply nightmarenet/utils/tracking.py:221]: Thanks for the thoughtful review! I’ve implemented the suggested changes and verified them locally. Appreciate the feedback.

**Review by Adit-Jain-srm — CHANGES_REQUESTED** [review]
LGTM. Please follow @Adit-Jain-srm - starred but follow not confirmed yet as per guidelines @Ayush-0918 .

**Comment by Ayush-0918** [comment]
@Adit-Jain-srm Done

**Review by Adit-Jain-srm — APPROVED** [review]
Great work!
